File: proxy-gateway-core/src/proxy_format.rs
//! Proxy line format parsing.
//!
//! Supports three common proxy line formats, all with an optional
//! `protocol://` prefix (which is stripped and ignored):
//!
//! - [`ProxyFormat::UserPassAtHostPort`] — `user:pass@host:port`
//! - [`ProxyFormat::UserPassHostPort`] — `user:pass:host:port`
//! - [`ProxyFormat::HostPortUserPass`] — `host:port:user:pass`
//!
//! Lines without credentials (`host:port`) are accepted by all formats.
//! IPv6 addresses in brackets (`[::1]:port`) are supported.

use crate::SourceProxy;
use anyhow::{Context, Result};

/// Supported proxy line formats.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProxyFormat {
    /// `username:password@host:port` (with optional `protocol://` prefix)
    UserPassAtHostPort,
    /// `username:password:host:port`
    UserPassHostPort,
    /// `host:port:username:password` (legacy default)
    HostPortUserPass,
}

impl Default for ProxyFormat {
    fn default() -> Self {
        Self::HostPortUserPass
    }
}

/// Parse a single proxy line in the given format.
///
/// An optional `protocol://` prefix (e.g. `http://`, `socks5://`) is stripped.
/// Lines without credentials (`host:port`) are accepted by all formats.
pub fn parse_proxy_line(s: &str, format: ProxyFormat) -> Result<SourceProxy> {
    // Strip optional protocol prefix.
    let s = strip_protocol(s);

    match format {
        ProxyFormat::UserPassAtHostPort => parse_user_pass_at_host_port(s),
        ProxyFormat::UserPassHostPort => parse_user_pass_host_port(s),
        ProxyFormat::HostPortUserPass => parse_host_port_user_pass(s),
    }
}

/// Strip an optional `protocol://` prefix.
fn strip_protocol(s: &str) -> &str {
    match s.find("://") {
        Some(i) => &s[i + 3..],
        None => s,
    }
}

// ---------------------------------------------------------------------------
// Format: user:pass@host:port
// ---------------------------------------------------------------------------

fn parse_user_pass_at_host_port(s: &str) -> Result<SourceProxy> {
    match s.rfind('@') {
        Some(at) => {
            let creds = &s[..at];
            let host_port = &s[at + 1..];
            let (host, port) = parse_host_port(host_port)?;
            let (user, pass) = split_user_pass(creds)?;
            Ok(SourceProxy {
                host,
                port,
                username: Some(user),
                password: Some(pass),
            })
        }
        None => {
            // No @ — treat as host:port without credentials.
            let (host, port) = parse_host_port(s)?;
            Ok(SourceProxy {
                host,
                port,
                username: None,
                password: None,
            })
        }
    }
}

// ---------------------------------------------------------------------------
// Format: user:pass:host:port
// ---------------------------------------------------------------------------

fn parse_user_pass_host_port(s: &str) -> Result<SourceProxy> {
    // Try 4-part split first (user:pass:host:port).
    // Ambiguity: we split from the right to find host:port, since the
    // password could contain colons in theory, but port is always numeric.
    let parts: Vec<&str> = s.rsplitn(3, ':').collect();
    // rsplitn(3, ':') on "u:p:h:port" gives ["port", "h", "u:p"]
    if parts.len() == 3 {
        if let Ok(port) = parts[0].parse::<u16>() {
            let host = parts[1];
            let creds = parts[2];
            if let Some(colon) = creds.find(':') {
                let user = &creds[..colon];
                let pass = &creds[colon + 1..];
                if !user.is_empty() {
                    return Ok(SourceProxy {
                        host: host.to_string(),
                        port,
                        username: Some(user.to_string()),
                        password: Some(pass.to_string()),
                    });
                }
            }
        }
    }

    // Fall back to host:port (no credentials).
    if parts.len() >= 2 {
        if let Ok(port) = parts[0].parse::<u16>() {
            // For 2-part: rsplitn gives ["port", "host"]
            // For 3-part where above failed: try treating it as host:port
            let host = if parts.len() == 2 {
                parts[1]
            } else {
                // Rejoin everything before the last colon
                let last_colon = s.rfind(':').unwrap();
                &s[..last_colon]
            };
            // Only accept as host:port if it looks like a host (no extra colons unless ipv6)
            if !host.contains(':') || host.starts_with('[') {
                return Ok(SourceProxy {
                    host: host.to_string(),
                    port,
                    username: None,
                    password: None,
                });
            }
        }
    }

    anyhow::bail!("expected user:pass:host:port or host:port, got '{s}'")
}

// ---------------------------------------------------------------------------
// Format: host:port:user:pass
// ---------------------------------------------------------------------------

fn parse_host_port_user_pass(s: &str) -> Result<SourceProxy> {
    // Handle IPv6 in brackets.
    if s.starts_with('[') {
        let bracket_end = s
            .find(']')
            .ok_or_else(|| anyhow::anyhow!("unclosed bracket in '{s}'"))?;
        let host = s[1..bracket_end].to_string();
        let rest = s[bracket_end + 1..]
            .strip_prefix(':')
            .ok_or_else(|| anyhow::anyhow!("expected ':' after ']' in '{s}'"))?;
        return parse_port_and_optional_creds(&host, rest);
    }

    let parts: Vec<&str> = s.splitn(4, ':').collect();
    match parts.len() {
        2 => {
            let port: u16 = parts[1].parse().context("invalid port")?;
            Ok(SourceProxy {
                host: parts[0].to_string(),
                port,
                username: None,
                password: None,
            })
        }
        4 => {
            let port: u16 = parts[1].parse().context("invalid port")?;
            Ok(SourceProxy {
                host: parts[0].to_string(),
                port,
                username: Some(parts[2].to_string()),
                password: Some(parts[3].to_string()),
            })
        }
        _ => anyhow::bail!("expected host:port or host:port:user:pass, got '{s}'"),
    }
}

fn parse_port_and_optional_creds(host: &str, rest: &str) -> Result<SourceProxy> {
    let parts: Vec<&str> = rest.splitn(3, ':').collect();
    match parts.len() {
        1 => {
            let port: u16 = parts[0].parse().context("invalid port")?;
            Ok(SourceProxy {
                host: host.to_string(),
                port,
                username: None,
                password: None,
            })
        }
        3 => {
            let port: u16 = parts[0].parse().context("invalid port")?;
            Ok(SourceProxy {
                host: host.to_string(),
                port,
                username: Some(parts[1].to_string()),
                password: Some(parts[2].to_string()),
            })
        }
        _ => anyhow::bail!("expected port or port:user:pass after host, got '{rest}'"),
    }
}

// ---------------------------------------------------------------------------
// Shared helpers
// ---------------------------------------------------------------------------

/// Parse `host:port` or `[ipv6]:port`.
fn parse_host_port(s: &str) -> Result<(String, u16)> {
    if s.starts_with('[') {
        let bracket_end = s
            .find(']')
            .ok_or_else(|| anyhow::anyhow!("unclosed bracket in '{s}'"))?;
        let host = s[1..bracket_end].to_string();
        let rest = s[bracket_end + 1..]
            .strip_prefix(':')
            .ok_or_else(|| anyhow::anyhow!("expected ':' after ']' in '{s}'"))?;
        let port: u16 = rest.parse().context("invalid port")?;
        Ok((host, port))
    } else {
        let colon = s
            .rfind(':')
            .ok_or_else(|| anyhow::anyhow!("expected host:port, got '{s}'"))?;
        let host = &s[..colon];
        let port: u16 = s[colon + 1..].parse().context("invalid port")?;
        Ok((host.to_string(), port))
    }
}

/// Split `user:pass` at the first colon.
fn split_user_pass(s: &str) -> Result<(String, String)> {
    let colon = s
        .find(':')
        .ok_or_else(|| anyhow::anyhow!("expected user:pass, got '{s}'"))?;
    Ok((s[..colon].to_string(), s[colon + 1..].to_string()))
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    // -- UserPassAtHostPort --

    #[test]
    fn test_upahp_basic() {
        let p = parse_proxy_line("user:pass@host.com:8080", ProxyFormat::UserPassAtHostPort).unwrap();
        assert_eq!(p.host, "host.com");
        assert_eq!(p.port, 8080);
        assert_eq!(p.username.as_deref(), Some("user"));
        assert_eq!(p.password.as_deref(), Some("pass"));
    }

    #[test]
    fn test_upahp_with_protocol() {
        let p = parse_proxy_line("http://user:pass@host.com:3128", ProxyFormat::UserPassAtHostPort).unwrap();
        assert_eq!(p.host, "host.com");
        assert_eq!(p.port, 3128);
        assert_eq!(p.username.as_deref(), Some("user"));
    }

    #[test]
    fn test_upahp_socks5() {
        let p = parse_proxy_line("socks5://user:pass@host.com:1080", ProxyFormat::UserPassAtHostPort).unwrap();
        assert_eq!(p.host, "host.com");
        assert_eq!(p.port, 1080);
    }

    #[test]
    fn test_upahp_no_creds() {
        let p = parse_proxy_line("host.com:8080", ProxyFormat::UserPassAtHostPort).unwrap();
        assert_eq!(p.host, "host.com");
        assert_eq!(p.port, 8080);
        assert!(p.username.is_none());
    }

    #[test]
    fn test_upahp_complex_username() {
        // Bottingtools-style username with underscores and hyphens
        let p = parse_proxy_line(
            "exampleuser_pool-custom_type-high:XpmTeTdYy8hT@proxy.bottingtools.com:1337",
            ProxyFormat::UserPassAtHostPort,
        ).unwrap();
        assert_eq!(p.host, "proxy.bottingtools.com");
        assert_eq!(p.port, 1337);
        assert_eq!(p.username.as_deref(), Some("exampleuser_pool-custom_type-high"));
        assert_eq!(p.password.as_deref(), Some("XpmTeTdYy8hT"));
    }

    #[test]
    fn test_upahp_ipv6() {
        let p = parse_proxy_line("user:pass@[::1]:3128", ProxyFormat::UserPassAtHostPort).unwrap();
        assert_eq!(p.host, "::1");
        assert_eq!(p.port, 3128);
        assert_eq!(p.username.as_deref(), Some("user"));
    }

    // -- UserPassHostPort --

    #[test]
    fn test_uphp_basic() {
        let p = parse_proxy_line("user:pass:host.com:8080", ProxyFormat::UserPassHostPort).unwrap();
        assert_eq!(p.host, "host.com");
        assert_eq!(p.port, 8080);
        assert_eq!(p.username.as_deref(), Some("user"));
        assert_eq!(p.password.as_deref(), Some("pass"));
    }

    #[test]
    fn test_uphp_no_creds() {
        let p = parse_proxy_line("host.com:8080", ProxyFormat::UserPassHostPort).unwrap();
        assert_eq!(p.host, "host.com");
        assert_eq!(p.port, 8080);
        assert!(p.username.is_none());
    }

    #[test]
    fn test_uphp_with_protocol() {
        let p = parse_proxy_line("http://user:pass:host.com:3128", ProxyFormat::UserPassHostPort).unwrap();
        assert_eq!(p.host, "host.com");
        assert_eq!(p.port, 3128);
        assert_eq!(p.username.as_deref(), Some("user"));
    }

    // -- HostPortUserPass --

    #[test]
    fn test_hpup_basic() {
        let p = parse_proxy_line("host.com:8080:user:pass", ProxyFormat::HostPortUserPass).unwrap();
        assert_eq!(p.host, "host.com");
        assert_eq!(p.port, 8080);
        assert_eq!(p.username.as_deref(), Some("user"));
        assert_eq!(p.password.as_deref(), Some("pass"));
    }

    #[test]
    fn test_hpup_no_creds() {
        let p = parse_proxy_line("host.com:8080", ProxyFormat::HostPortUserPass).unwrap();
        assert_eq!(p.host, "host.com");
        assert_eq!(p.port, 8080);
        assert!(p.username.is_none());
    }

    #[test]
    fn test_hpup_ipv6() {
        let p = parse_proxy_line("[::1]:3128", ProxyFormat::HostPortUserPass).unwrap();
        assert_eq!(p.host, "::1");
        assert_eq!(p.port, 3128);
    }

    #[test]
    fn test_hpup_ipv6_with_creds() {
        let p = parse_proxy_line("[2001:db8::1]:8080:user:pass", ProxyFormat::HostPortUserPass).unwrap();
        assert_eq!(p.host, "2001:db8::1");
        assert_eq!(p.port, 8080);
        assert_eq!(p.username.as_deref(), Some("user"));
    }

    #[test]
    fn test_hpup_with_protocol() {
        let p = parse_proxy_line("http://host.com:8080:user:pass", ProxyFormat::HostPortUserPass).unwrap();
        assert_eq!(p.host, "host.com");
        assert_eq!(p.port, 8080);
        assert_eq!(p.username.as_deref(), Some("user"));
    }

    #[test]
    fn test_hpup_bad_format() {
        assert!(parse_proxy_line("host:port:only_three", ProxyFormat::HostPortUserPass).is_err());
        assert!(parse_proxy_line("justhost", ProxyFormat::HostPortUserPass).is_err());
    }
}

File: proxy-gateway/core/proxy_format.go
package core

import (
	"fmt"
	"strconv"
	"strings"
)

// ProxyFormat describes how proxy lines are formatted in a file.
type ProxyFormat string

const (
	// ProxyFormatHostPortUserPass: host:port:user:pass (default)
	ProxyFormatHostPortUserPass ProxyFormat = "host_port_user_pass"
	// ProxyFormatUserPassAtHostPort: user:pass@host:port
	ProxyFormatUserPassAtHostPort ProxyFormat = "user_pass_at_host_port"
	// ProxyFormatUserPassHostPort: user:pass:host:port
	ProxyFormatUserPassHostPort ProxyFormat = "user_pass_host_port"
)

// DefaultProxyFormat is the default format.
const DefaultProxyFormat = ProxyFormatHostPortUserPass

// ParseProxyLine parses a single proxy line in the given format.
// An optional protocol:// prefix is stripped. Lines without credentials (host:port) are accepted.
func ParseProxyLine(s string, format ProxyFormat) (SourceProxy, error) {
	s = stripProtocol(s)
	switch format {
	case ProxyFormatUserPassAtHostPort:
		return parseUserPassAtHostPort(s)
	case ProxyFormatUserPassHostPort:
		return parseUserPassHostPort(s)
	default: // host_port_user_pass
		return parseHostPortUserPass(s)
	}
}

func stripProtocol(s string) string {
	if i := strings.Index(s, "://"); i >= 0 {
		return s[i+3:]
	}
	return s
}

func parseUserPassAtHostPort(s string) (SourceProxy, error) {
	at := strings.LastIndex(s, "@")
	if at < 0 {
		host, port, err := parseHostPort(s)
		if err != nil {
			return SourceProxy{}, err
		}
		return SourceProxy{Host: host, Port: port}, nil
	}
	creds := s[:at]
	hostPort := s[at+1:]
	host, port, err := parseHostPort(hostPort)
	if err != nil {
		return SourceProxy{}, err
	}
	user, pass, err := splitUserPass(creds)
	if err != nil {
		return SourceProxy{}, err
	}
	return SourceProxy{Host: host, Port: port, Username: &user, Password: &pass}, nil
}

func parseUserPassHostPort(s string) (SourceProxy, error) {
	// Split from the right into at most 3 parts: [creds, host, port]
	// For "user:pass:host.com:8080" → ["user:pass", "host.com", "8080"]
	parts := rSplitN(s, ':', 3)
	if len(parts) == 3 {
		// parts[0]=creds, parts[1]=host, parts[2]=port
		if port, err := strconv.ParseUint(parts[2], 10, 16); err == nil {
			host := parts[1]
			creds := parts[0]
			if ci := strings.Index(creds, ":"); ci >= 0 {
				user := creds[:ci]
				pass := creds[ci+1:]
				if user != "" {
					p := uint16(port)
					return SourceProxy{Host: host, Port: p, Username: &user, Password: &pass}, nil
				}
			}
		}
	}
	// Fall back to host:port
	host, port, err := parseHostPort(s)
	if err != nil {
		return SourceProxy{}, fmt.Errorf("expected user:pass:host:port or host:port, got %q", s)
	}
	return SourceProxy{Host: host, Port: port}, nil
}

func parseHostPortUserPass(s string) (SourceProxy, error) {
	// Handle IPv6 in brackets
	if strings.HasPrefix(s, "[") {
		end := strings.Index(s, "]")
		if end < 0 {
			return SourceProxy{}, fmt.Errorf("unclosed bracket in %q", s)
		}
		host := s[1:end]
		rest := s[end+1:]
		if !strings.HasPrefix(rest, ":") {
			return SourceProxy{}, fmt.Errorf("expected ':' after ']' in %q", s)
		}
		return parsePortAndOptionalCreds(host, rest[1:])
	}

	parts := strings.SplitN(s, ":", 4)
	switch len(parts) {
	case 2:
		port, err := strconv.ParseUint(parts[1], 10, 16)
		if err != nil {
			return SourceProxy{}, fmt.Errorf("invalid port in %q", s)
		}
		return SourceProxy{Host: parts[0], Port: uint16(port)}, nil
	case 4:
		port, err := strconv.ParseUint(parts[1], 10, 16)
		if err != nil {
			return SourceProxy{}, fmt.Errorf("invalid port in %q", s)
		}
		user, pass := parts[2], parts[3]
		return SourceProxy{Host: parts[0], Port: uint16(port), Username: &user, Password: &pass}, nil
	default:
		return SourceProxy{}, fmt.Errorf("expected host:port or host:port:user:pass, got %q", s)
	}
}

func parsePortAndOptionalCreds(host, rest string) (SourceProxy, error) {
	parts := strings.SplitN(rest, ":", 3)
	switch len(parts) {
	case 1:
		port, err := strconv.ParseUint(parts[0], 10, 16)
		if err != nil {
			return SourceProxy{}, fmt.Errorf("invalid port %q", parts[0])
		}
		return SourceProxy{Host: host, Port: uint16(port)}, nil
	case 3:
		port, err := strconv.ParseUint(parts[0], 10, 16)
		if err != nil {
			return SourceProxy{}, fmt.Errorf("invalid port %q", parts[0])
		}
		user, pass := parts[1], parts[2]
		return SourceProxy{Host: host, Port: uint16(port), Username: &user, Password: &pass}, nil
	default:
		return SourceProxy{}, fmt.Errorf("expected port or port:user:pass after host, got %q", rest)
	}
}

func parseHostPort(s string) (string, uint16, error) {
	if strings.HasPrefix(s, "[") {
		end := strings.Index(s, "]")
		if end < 0 {
			return "", 0, fmt.Errorf("unclosed bracket in %q", s)
		}
		host := s[1:end]
		rest := s[end+1:]
		if !strings.HasPrefix(rest, ":") {
			return "", 0, fmt.Errorf("expected ':' after ']' in %q", s)
		}
		port, err := strconv.ParseUint(rest[1:], 10, 16)
		if err != nil {
			return "", 0, fmt.Errorf("invalid port in %q", s)
		}
		return host, uint16(port), nil
	}
	i := strings.LastIndex(s, ":")
	if i < 0 {
		return "", 0, fmt.Errorf("expected host:port, got %q", s)
	}
	port, err := strconv.ParseUint(s[i+1:], 10, 16)
	if err != nil {
		return "", 0, fmt.Errorf("invalid port in %q", s)
	}
	return s[:i], uint16(port), nil
}

func splitUserPass(s string) (string, string, error) {
	i := strings.Index(s, ":")
	if i < 0 {
		return "", "", fmt.Errorf("expected user:pass, got %q", s)
	}
	return s[:i], s[i+1:], nil
}

// rSplitN splits s at c from the right, returning at most n parts.
func rSplitN(s string, c byte, n int) []string {
	parts := make([]string, 0, n)
	remaining := s
	for len(parts) < n-1 {
		i := strings.LastIndexByte(remaining, c)
		if i < 0 {
			break
		}
		parts = append(parts, remaining[i+1:])
		remaining = remaining[:i]
	}
	parts = append(parts, remaining)
	// reverse
	for i, j := 0, len(parts)-1; i < j; i, j = i+1, j-1 {
		parts[i], parts[j] = parts[j], parts[i]
	}
	return parts
}
